Issue #78 addition of Correlation ID to specification

SensibleWood opened the pull request:
Fixes #78 

As follows:

* Added Correlation ID object
* Added runtime expression support
* Added Correlation ID Map to Components Object
* Added Correlation ID to Message Object
* Updated schema and added example to test

A couple of bits of food for thought:

* Is the ABNF definition too restrictive?
* Is it right to match the ABNF definition in schema?
* **Would welcome suggestions as to whether more references are needed anywhere else in the specification as it stands i.e. defining a Correlation ID at a higher level than Message**

fmvilas commented:
Added "Fixes #78" so the issue is closed when the PR gets merged.

fmvilas commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+A runtime expression allows values to be defined based on information that will be available within the message.
+This mechanism is used by [Correlation ID Object](#correlationIdObject).
+
+The runtime expression is defined by the following [ABNF](https://tools.ietf.org/html/rfc5234) syntax
```
```suggestion
The runtime expression is defined by the following [ABNF](https://tools.ietf.org/html/rfc5234) syntax:
```

fmvilas commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+      source = ( header-reference | payload-reference )  
+      header-reference = "header." token
+      payload-reference = "payload." name
+      fragment = a JSON Pointer [RFC 6901](https://tools.ietf.org/html/rfc6901) 
```
What's fragment doing here?

SensibleWood replied:
Fragment is the last piece of the ABNF notation that points - in JSON Path syntax - to the specific property. So it's the `.correlationId` in `$message.header.correlationId`

SensibleWood replied:
There's a missing piece from the port from OpenAPI I need to add back in - will update the PR with changes

fmvilas commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+      header-reference = "header." token
+      payload-reference = "payload." name
+      fragment = a JSON Pointer [RFC 6901](https://tools.ietf.org/html/rfc6901) 
+      name = *( char )
```
If `name` allows for "zero" chars, we could end up having something like `$message.payload.`.

What about this?
```suggestion
      name = 1*( char )
```

SensibleWood replied:
Works for me

fmvilas commented on versions/next/asyncapi.md:
````diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+      token = as per RFC [7230](https://tools.ietf.org/html/rfc7230#section-3.2.6)
+```
+
+The `name` identifier is case-sensitive, whereas `token` is not. 
````
Why?

SensibleWood replied:
On review I need to remove the references to `token` - will change and update PR

fmvilas commented on versions/next/asyncapi.md:
````diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+The runtime expression is defined by the following [ABNF](https://tools.ietf.org/html/rfc5234) syntax
+
+```
+      expression = ( "$message" )
````
I think there's something missing here:

```suggestion
      expression = ( "$message" source )
```

fmvilas commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+
+Source Location | Example expression  | Notes
+---|:---|:---|
+Message Header Property | `$message.header.correlationId` | Headers are analogous to message properties
```
```suggestion
Message Header Property | `$message.header.correlationId` | Headers are analogous to message headers.
```

SensibleWood replied:
What I was getting here is that some messaging providers call headers properties, so giving some guidance for the reader if they understand the meaning of properties not properties

fmvilas replied:
Yeah, but that should be explained in the message headers section, don't you think so? This note seems confusing to me because we never called them `message properties` (on purpose) and suddenly it appears here.

SensibleWood replied:
Sure sounds good - will add something appropriate and then change the wording here

SensibleWood replied:
Have changed the examples and hopefully made them more useful

fmvilas replied:
Indeed. Thanks!

fmvilas commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+Source Location | Example expression  | Notes
+---|:---|:---|
+Message Header Property | `$message.header.correlationId` | Headers are analogous to message properties
+Message Payload Property | `$message.payload.meta.messageId` | Correlation ID is set using a property from the `meta` section of a message
```
```suggestion
Message Payload Property | `$message.payload.meta.messageId` | Correlation ID is set using a property from the `meta` section of a message.
```

fmvilas commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+Message Header Property | `$message.header.correlationId` | Headers are analogous to message properties
+Message Payload Property | `$message.payload.meta.messageId` | Correlation ID is set using a property from the `meta` section of a message
+
+Runtime expressions preserve the type of the referenced value.
```
I think this answers my suggestion about getting rid of `type` in correlation Ids, right?

SensibleWood replied:
Yes it does - going to remove and update PR

fmvilas commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+Message Payload Property | `$message.payload.meta.messageId` | Correlation ID is set using a property from the `meta` section of a message
+
+Runtime expressions preserve the type of the referenced value.
+Expressions can be embedded into string values by surrounding the expression with `{}` curly braces.
```
Do you have an example where this should be possible?

SensibleWood replied:
Where you use a custom correlation ID? This would be down to API designers - if they want to enforce a correlation ID and cannot add a message header property then this offers the requisite features to use something in the payload. Alternatively, they may be retrofitting an API specification to a messaging platform and already use a custom correlation ID in the message payload.

Is this what you were getting at?

SensibleWood replied:
You mean the curly braces? Will add an example

fmvilas replied:
I'm more worried about where you can use runtime expressions. Specification-wise it's easy to say we support them, but tooling-wise it's not that simple. I want to make sure we only support things where we really need them.

SensibleWood replied:
On reflection we don't support this at the moment. In OpenAPI it's a construct that relates to [Callback Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#callbackObject). I can't see a good fit so will remove the guidance.

SensibleWood replied:
> I'm more worried about where you can use runtime expressions. Specification-wise it's easy to say we support them, but tooling-wise it's not that simple. I want to make sure we only support things where we really need them.

I think the runtime expression support is a really good investment for the future. If you look at the way Callback Object in OpenAPI can now be engineered it adds a great deal of flexibility for the API designer. Specifically, when we tackle RPC in earnest a runtime expression to specify replyTo queues and the like based on attributes of the message will give API designers a real boost IMHO.

fmvilas replied:
Yeah, I get it. But I still can't see the value outside the correlation ID feature or a possible RPC support in the future. My concern is just that this sentence really looked like you can embed runtime expressions anywhere you have a string. Anyway, I think your previous comment clarified it :)

SensibleWood replied:
👍

SensibleWood commented on examples/next/correlation-id.yml:
```diff
@@ -0,0 +1,106 @@
+      title: Light measured
+      summary: Inform about environmental lighting conditions for a particular streetlight.
+      correlationId:
+        location: "$message.header.correlationId"
```
`.header.` is the way this is specified in OpenAPI:

![image](https://user-images.githubusercontent.com/2420069/53816943-6f65e280-3f5c-11e9-9c4c-c3a38eac728b.png)

Personally I think we should be in sync.

fmvilas replied:
You're right, we should be in sync. I didn't know OpenAPI was using `header`. Thanks!

SensibleWood commented on examples/next/correlation-id.yml:
```diff
@@ -0,0 +1,106 @@
+  correlationIds:
+    sentAtCorrelator:
+      description: Data from message payload used as correlation ID
+      location: $message.payload.sentAt
```
`sentAt` is an attribute of the messages in the example:

```  dimLightPayload:
      type: object
      properties:
        percentage:
          type: integer
          description: Percentage to which the light should be dimmed to.
          minimum: 0
          maximum: 100
        sentAt:
          $ref: "#/components/schemas/sentAt"```

The idea of the example is to generalise the reference i.e. using the ABNF you don't need to have it specifically wedded to a schema. However, if it's confusing then I guess it needs to change. Will sort out a different approach.

fmvilas replied:
Oh! I thought you forgot it here but now I see where it's being applied. In this case, it's fine. Thanks for clarifying!

SensibleWood replied:
OK no worries - will leave as is then

SensibleWood commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+description | `string` | A optional description of the identifier. [CommonMark syntax](http://spec.commonmark.org/) can be used for rich text representation.
+type | `integer` \| `string` \| `number` | An optional type specification.
+location | {expression} | **REQUIRED.** A runtime expression that specifies the location of the correlation ID.
+
```
OK sounds good - will add

SensibleWood commented on versions/next/asyncapi.md:
```diff
@@ -2208,6 +2231,68 @@ petstore_auth:
+Field Name | Type | Description
+---|:---|---
+description | `string` | A optional description of the identifier. [CommonMark syntax](http://spec.commonmark.org/) can be used for rich text representation.
+type | `integer` \| `string` \| `number` | An optional type specification.
```
I was in two minds whether we need `type` or not - the design of this feature has moved on - so I'm going to remove it as it seems overkill.

SensibleWood commented:
Made changes as requested. Please note that the ABNF syntax was incorrect and I've changed as follows:

* Removed `name`, `char` and `token`
* Both `header` and `payload` are now reliant on `fragment`

I think this fits better for headers as they can be nested e.g. if you look at WebSphere MQ they support (for example) CorrelationId inside the header MQMD - so using a JSON Pointer to provide a link to value seems like the right approach.
